app/utils/sql/SQLDAO.scala

```diff
@@ -0,0 +1,99 @@
+import utils.ObjectId
+
+import javax.inject.Inject
+import scala.annotation.nowarn
+import scala.concurrent.ExecutionContext
+
+import slick.jdbc.PostgresProfile.api._
+
+abstract class SQLDAO[C, R, X <: AbstractTable[R]] @Inject()(sqlClient: SqlClient)(implicit ec: ExecutionContext)
+    extends SecuredSQLDAO(sqlClient) {
+  protected def collection: TableQuery[X]
+  override protected def collectionName: String =
+    collection.shaped.value.schemaName.map(_ + ".").getOrElse("") + collection.shaped.value.tableName
+
+  protected def columnsList: List[String] = collection.baseTableRow.create_*.map(_.name).toList
+  def columns: SqlToken = SqlToken.raw(columnsList.mkString(", "))
+  def columnsWithPrefix(prefix: String): SqlToken = SqlToken.raw(columnsList.map(prefix + _).mkString(", "))
+
+  protected def idColumn(x: X): Rep[String]
+  protected def isDeletedColumn(x: X): Rep[Boolean]
+
+  protected def notdel(r: X): Rep[Boolean] = isDeletedColumn(r) === false
+
+  protected def parse(row: X#TableElementType): Fox[C]
+
+  protected def parseFirst(rowSeq: Seq[X#TableElementType], queryLabel: ObjectId): Fox[C] =
+    parseFirst(rowSeq, queryLabel.toString)
+
+  protected def parseFirst(rowSeq: Seq[X#TableElementType], queryLabel: String): Fox[C] =
+    for {
+      firstRow <- rowSeq.headOption.toFox // No error chain here, as this should stay Fox.Empty
+      parsed <- parse(firstRow) ?~> s"Parsing failed for row in $collectionName queried by $queryLabel"
+    } yield parsed
+
+  protected def parseAll(rowSeq: Seq[X#TableElementType]): Fox[List[C]] =
+    Fox.combined(rowSeq.toList.map(parse)) ?~> s"Parsing failed for a row in $collectionName during list query"
+
+  @nowarn // suppress warning about unused implicit ctx, as it is used in subclasses
+  def findOne(id: ObjectId)(implicit ctx: DBAccessContext): Fox[C] =
+    run(collection.filter(r => isDeletedColumn(r) === false && idColumn(r) === id.id).result.headOption).map {
+      case Some(r) =>
+        parse(r) ?~> ("sql: could not parse database row for object" + id)
+      case _ =>
+        Fox.failure("sql: could not find object " + id)
+    }.flatten
+
+  @nowarn // suppress warning about unused implicit ctx, as it is used in subclasses
+  def findAll(implicit ctx: DBAccessContext): Fox[List[C]] =
+    for {
+      r <- run(collection.filter(row => notdel(row)).result)
+      parsed <- Fox.combined(r.toList.map(parse))
+    } yield parsed
+
+  def deleteOne(id: ObjectId)(implicit ctx: DBAccessContext): Fox[Unit] = {
+    val q = for { row <- collection if notdel(row) && idColumn(row) === id.id } yield isDeletedColumn(row)
+    for {
+      _ <- assertDeleteAccess(id)
+      _ <- run(q.update(true))
+    } yield ()
+  }
+
+  protected def updateStringCol(id: ObjectId, column: X => Rep[String], newValue: String)(
+      implicit ctx: DBAccessContext): Fox[Unit] = {
+    val q = for { row <- collection if notdel(row) && idColumn(row) === id.id } yield column(row)
```

Comment:
`val q` shadows the `q`-query builder

Reply:
Good catch, I renamed these to query. I am thinking of removing those entirely in the near future, though, to have only one unified way of creating sql queries.